File: ruby/grains/grains.rb
class Grains
  class << self
    def squares(n)
      [*0..(n-1)].map{|num| 2 ** num}
    end
    def square(n)
      raise ArgumentError if n <= 0 || 64 < n
      squares(n)[n-1]
    end
    def total
      squares(64).sum
    end
  end
end

File: haskell/grains/src/Grains.hs
module Grains (square, total) where

import Data.Maybe (mapMaybe)

squareLimit :: Integer
squareLimit = 64

square :: Integer -> Maybe Integer
square n
  | n >= 1 && n <= squareLimit = Just (2 ^ (n - 1))
  | otherwise = Nothing

total :: Integer
total = sum $ mapMaybe square [1 .. squareLimit]
